### ISSUE TITLE
Validate optional GIoU loss against an independent oracle

### ISSUE THREAD

**Issue by shapovalovei** [shapovalovei/YOLOX-OBB#50]
## Phase

Bounded validation only. Implementation is NOT authorized by this Issue.

## Repository baseline

Repository: `shapovalovei/YOLOX-OBB`

Current main: `6ddaed5d5d6103b1de11d14cf3e65167b84c8c73`

Behavioral baseline: `0695e2834ff5716dfd51fcd6ac7b2300b7bc27ce`

Validation must preserve the maintained OBB contract and must not change tracked files.

## Global constraints

- no source, test, configuration, or documentation changes;
- no branch, commit, PR, or CI;
- no training, optimizer steps, dataset changes, or model-quality claims;
- temporary probes/checkpoints may exist only outside the repository;
- separate FACT, INFERENCE, and UNVERIFIED;
- owner/orchestrator reviews and closes this Issue.

## Local hypothesis

The optional `IOUloss(loss_type="giou")` path in `yolox/models/losses.py` appears to use the intersection area in the generalized-area penalty where the union area is required.

A local probe with prediction `[0, 0, 2, 2]` and target `[1, 1, 2, 2]` returned `1.7460318`. An independently calculated generalized-IoU loss for the same valid boxes is `1.0793651`. The default local loss path uses IoU, so exposure depends on configuration.

## Validation matrix

Build an independent mathematical oracle in the temporary probe. Do not copy implementation code or formulas from another repository.

For valid positive xyxy boxes, cover identical, partial-overlap, containment, edge-contact, disjoint, unequal-size, and batched cases. Compare local forward values at float32 and float64 where supported. Compare prediction gradients using autograd plus finite-difference or high-precision checks. Verify the default IoU path is unchanged.

## Acceptance / falsification

**A — acceptable:** GIoU forward values and gradients match the independent oracle within reported tolerances, and IoU behavior is unchanged.

**B — defect proven:** any valid case shows a formula-level forward or gradient mismatch beyond numerical tolerance.

**C — inconclusive:** the optional path cannot be reached or an independent oracle cannot be evaluated reliably.

## Stop conditions

No loss rewrite, OBB KLD change, training, model-quality evaluation, tracked test, or implementation is authorized.

## Completion handoff

Post the oracle definition, cases, tolerances, values, gradients, environment, FACT / INFERENCE / UNVERIFIED, verdict A/B/C, and smallest future implementation scope if B is proven.

**Comment by shapovalovei** [comment]
Orchestrator baseline update before validation: current `origin/main` is `6c0d769b1fce70a314b1aff5d690dc86dc91d120`. The hypothesis/scope of #50 are unchanged. Use this current main for preflight and validation; the older SHA in the Issue body is historical.

**Comment by shapovalovei** [comment]
## Validation handoff

### Baseline and environment

- Repository: `shapovalovei/YOLOX-OBB`
- Validation baseline: `main` at `6c0d769b1fce70a314b1aff5d690dc86dc91d120`
- `origin/main`: `6c0d769b1fce70a314b1aff5d690dc86dc91d120`
- Required preflight completed: clean worktree; `git fetch origin`; `git switch main`; `git pull --ff-only` (already up to date).
- Final worktree check: clean, still on `main`, local `HEAD == origin/main`.
- Maintained path exercised: `from yolox.models.losses import IOUloss`, `IOUloss(reduction="none", loss_type="giou")`.
- Runtime: `/Users/eugene/creditCardReader/venv/bin/python` 3.11.11; PyTorch 2.2.2; mpmath 1.3.0; macOS 26.6.2 x86_64.
- Probe was an inline temporary script outside the repository. No tracked files, branch, commit, PR, training, or CI were created.

The older SHA in the Issue body was treated as historical, per the latest orchestrator baseline comment; validation used the current SHA above.

### Independent oracle and equations

The oracle was derived directly from the mathematical GIoU definition and did not import or copy a formula/implementation from another repository. Inputs are the maintained `(cx, cy, w, h)` representation with positive `w,h`; the oracle first maps each box to corners:

```text
x_min = cx - w/2       y_min = cy - h/2
x_max = cx + w/2       y_max = cy + h/2
```

For prediction `P` and target `T`:

```text
iw = max(0, min(P.x_max, T.x_max) - max(P.x_min, T.x_min))
ih = max(0, min(P.y_max, T.y_max) - max(P.y_min, T.y_min))
I  = iw * ih

A_P = (P.x_max - P.x_min) * (P.y_max - P.y_min)
A_T = (T.x_max - T.x_min) * (T.y_max - T.y_min)
U  = A_P + A_T - I

C = (max(P.x_max, T.x_max) - min(P.x_min, T.x_min))
     * (max(P.y_max, T.y_max) - min(P.y_min, T.y_min))

IoU  = I / U
GIoU = IoU - (C - U) / C
loss = 1 - GIoU
```

The maintained default IoU convention was independently evaluated as `1 - IoU**2`, matching the `loss_type="iou"` branch. The oracle used exact positive-box areas; the maintained implementation's `+1e-16` denominator safeguard is immaterial at the reported tolerances.

### Forward case matrix

`P` and `T` below are `(cx, cy, w, h)`. `I/U/C` are oracle areas. Forward errors are absolute and relative to the oracle loss. Float64 is the primary mathematical comparison; float32 is the practical maintained path.

| case | P | T | I | U | C | oracle loss | maintained f64 (abs, rel) | maintained f32 (abs, rel) |
|---|---|---|---:|---:|---:|---:|---:|---:|
| identical | `[0,0,4,2]` | `[0,0,4,2]` | 8 | 8 | 8 | 0 | 0 (`0`, `0`) | 0 (`0`, `0`) |
| partial overlap | `[0,0,4,2]` | `[0.7,0.3,3,1.2]` | 3.36 | 8.24 | 8.4 | 0.611280629 | 1.192233010 (`5.810e-1`, `9.504e-1`) | 1.192232966 (`5.810e-1`, `9.504e-1`) |
| containment | `[0,0,6,4]` | `[0.4,0.3,2,1.2]` | 2.4 | 24 | 24 | 0.9 | 1.8 (`9.000e-1`, `1.000e+0`) | 1.799999952 (`9.000e-1`, `1.000e+0`) |
| edge contact | `[0,0,2,2]` | `[2,0.2,2,1.2]` | 0 | 6.4 | 8 | 1.2 | 2 (`8.000e-1`, `6.667e-1`) | 2 (`8.000e-1`, `6.667e-1`) |
| fully disjoint | `[-1,-1,2,2]` | `[4,3,1,1]` | 0 | 5 | 35.75 | 1.860139860 | 2 (`1.399e-1`, `7.519e-2`) | 2 (`1.399e-1`, `7.519e-2`) |
| unequal sizes | `[1,-0.5,7,1.4]` | `[0.2,0.3,1.1,3.2]` | 1.54 | 11.78 | 22.4 | 1.343377092 | 1.800519949 (`4.571e-1`, `3.403e-1`) | 1.800519943 (`4.571e-1`, `3.403e-1`) |
| strongly different aspect ratios | `[0,0,12,0.6]` | `[0.3,0.2,0.8,8]` | 0.48 | 13.12 | 96 | 1.826747967 | 1.958414634 (`1.317e-1`, `7.208e-2`) | 1.958414555 (`1.317e-1`, `7.208e-2`) |
| Issue probe | `[0,0,2,2]` | `[1,1,2,2]` | 1 | 7 | 9 | 1.079365079 | 1.746031746 (`6.667e-1`, `6.176e-1`) | 1.746031761 (`6.667e-1`, `6.176e-1`) |

The Issue probe reproduces the reported maintained float32 value (`1.7460318` to the Issue's precision) versus the oracle (`1.0793651`).

Batched mixed input used the partial-overlap, containment, edge-contact, fully-disjoint, and aspect-ratio rows in that order. Maintained output was:

```text
float64: [1.192233010, 1.800000000, 2.000000000, 2.000000000, 1.958414634]
float32: [1.192232966, 1.799999952, 2.000000000, 2.000000000, 1.958414555]
oracle:   [0.611280629, 0.900000000, 1.200000000, 1.860139860, 1.826747967]
```

Maximum GIoU forward error across the individual matrix was `0.900000` absolute / `1.000000` relative for float64 and `0.899999952` / `0.999999947` for float32. The batched maximum errors were the same within dtype rounding.

### Gradient validation

For multiple valid, overlapping, non-degenerate cases, maintained autograd was compared with a central finite-difference gradient of the independent oracle using mpmath at 80 decimal digits and `h = 1e-20`. Repeating with `h = 1e-25` changed the oracle gradients by at most `4.3e-42` across these cases.

| case | independent oracle gradient | maintained f64 autograd; abs / rel error | maintained f32 autograd; max abs / rel error |
|---|---|---|---|
| partial non-degenerate | `[0, -0.4537344320, 0.0881565991, -0.1424754666]` | `[0, -0.6382050203, 0.1365995056, -0.1208284078]`; max `1.845e-1` / `5.495e-1` | `[0, -0.6382049918, 0.1365994960, -0.1208283901]`; `1.845e-1` / `5.495e-1` |
| containment non-degenerate | `[0, 0, 0.0160416667, 0.0240625000]` | `[0, 0, 0.0320833333, 0.0481250000]`; max `2.406e-2` / `1.000e+0` | `[0, 0, 0.0320833325, 0.0481249990]`; `2.406e-2` / `1.000e+0` |
| aspect-ratio non-degenerate | `[0, 0, 0.0347776615, -0.2423005165]` | `[0, 0, 0.0211114807, -0.0807358907]`; max `1.616e-1` / `6.668e-1` | `[0, 0, 0.0211114790, -0.0807358995]`; `1.616e-1` / `6.668e-1` |

Screening tolerances were `1e-12` absolute/relative for float64 forward values, `1e-6` for float32 forward values, `1e-10` absolute / `1e-8` relative for float64 gradients, and `1e-5` absolute / `1e-4` relative for float32 gradients. The observed formula errors are orders of magnitude larger.

### Suspected penalty boundary

The maintained source at `yolox/models/losses.py:44` is:

```python
giou = iou - (area_c - area_i) / area_c.clamp(1e-16)
```

The source's `area_i` is intersection area. The GIoU definition requires the union area `U = area_p + area_g - area_i` in the generalized-area penalty. Direct decomposition proves the difference:

| case | I | U | C | maintained `(C-I)/C` | oracle `(C-U)/C` | delta |
|---|---:|---:|---:|---:|---:|---:|
| partial overlap | 3.36 | 8.24 | 8.4 | 0.6 | 0.019047619 | 0.580952381 |
| containment | 2.4 | 24 | 24 | 0.9 | 0 | 0.9 |
| edge contact | 0 | 6.4 | 8 | 1 | 0.2 | 0.8 |
| fully disjoint | 0 | 5 | 35.75 | 1 | 0.860139860 | 0.139860140 |
| Issue probe | 1 | 7 | 9 | 0.888888889 | 0.222222222 | 0.666666667 |

Thus the Issue hypothesis is proven mathematically. The enclosing rectangle itself is correct in these probes; the smallest causal implementation boundary is the generalized-area term in the optional `loss_type == "giou"` branch at line 44, specifically the use of `area_i` where the union quantity is required. No implementation was made.

### Default IoU path

The `loss_type="iou"` branch remains separate and is the current `1 - IoU**2` convention. Against the independent `1 - (I/U)**2` oracle over the same cases, float64 maximum absolute/relative error was `0 / 0`; float32 maximum was `3.316e-8` / `3.977e-8`. The maintained `yolo_head.py` call site constructs `IOUloss(reduction="none")` without selecting GIoU.

### Reachability

Tracked local source/config/experiment search (`git grep` for `giou`, `generalized.?iou`, `loss_type`, and `IOUloss`) found no active configuration or experiment selecting `loss_type="giou"`. The only maintained construction is the default IoU path in `yolox/models/yolo_head.py`; the GIoU branch remains directly reachable as an optional constructor argument. Reachability classification: **optional but reachable; apparently unused by current maintained configs/experiments**. This affects exposure/severity, not mathematical correctness.

### FACT / INFERENCE / UNVERIFIED

- **FACT:** On the stated baseline, valid positive boxes produce forward and prediction-gradient mismatches between maintained `loss_type="giou"` and the independent GIoU oracle, well beyond dtype/numerical tolerances.
- **FACT:** The mismatch is caused by `area_i` rather than union area in the optional GIoU penalty at `yolox/models/losses.py:44`.
- **FACT:** The default IoU path matches its current convention within float64/float32 numerical error and is source-separated from GIoU.
- **FACT:** No local tracked config/experiment selects GIoU; direct constructor use remains reachable.
- **INFERENCE:** The defect is isolated to optional generic GIoU and does not imply an OBB/KLD or rotated-geometry defect.
- **UNVERIFIED:** No claim is made about downstream training impact, model quality, or external consumers that are not represented in this checkout.

## Verdict

**B — DEFECT PROVEN.**

Recommended next phase: a separately authorized implementation Issue for the one-line optional generic GIoU penalty correction, followed by focused forward/gradient regressions covering the cases above and maintained IoU compatibility checks. Do not change OBB/KLD or rotated geometry as part of that follow-up. This validation Issue remains open for orchestrator review.

**Comment by shapovalovei** [comment]
Orchestrator validation review accepted.

FACT: On baseline `6c0d769b1fce70a314b1aff5d690dc86dc91d120`, `IOUloss(loss_type="giou")` computes IoU using union `U = area_p + area_g - area_i`, but the generalized-area penalty is `(area_c - area_i) / area_c` instead of `(area_c - U) / area_c`. This is a formula-level GIoU defect, not a numerical tolerance issue.

The independent oracle matrix and gradient checks in comment #5492699310 demonstrate large forward/gradient mismatches on valid non-degenerate boxes. The default `loss_type="iou"` branch matches its maintained `1 - IoU**2` convention and is source-separated from the defect. Local tracked configs/experiments do not select GIoU, so current exposure is low, but the optional constructor path remains reachable.

Verdict B is accepted. Smallest causal boundary: the optional generic GIoU penalty in `yolox/models/losses.py`. No OBB/KLD/rotated-geometry implication is inferred. Proceed with a separate implementation Issue; this validation Issue may be closed Completed.
